lib/media/track/localaudiotrack.js

```diff
@@ -20,14 +22,114 @@ const LocalMediaAudioTrack = mixinLocalMediaTrack(AudioTrack);
  * @emits LocalAudioTrack#started
  * @emits LocalAudioTrack#stopped
  */
+
 class LocalAudioTrack extends LocalMediaAudioTrack {
   /**
    * Construct a {@link LocalAudioTrack} from a MediaStreamTrack.
    * @param {MediaStreamTrack} mediaStreamTrack - An audio MediaStreamTrack
    * @param {LocalTrackOptions} [options] - {@link LocalTrack} options
    */
   constructor(mediaStreamTrack, options) {
+    // Note(mpatwardhan): by default workaround for WebKitBug1208516 will be enabled on Safari browsers
+    // although the bug is seen  mainly on iOS devices, we do not have a reliable way to tell iOS from MacOs
+    // userAgent on iOS pretends its macOs if Safari is set to request desktop pages.
+    const workaroundWebKitBug1208516 = guessBrowser() === 'safari';
+
+    options = Object.assign({
+      workaroundWebKitBug1208516,
+      getUserMedia,
+    }, options);
+
     super(mediaStreamTrack, options);
+    Object.defineProperties(this, {
+      _watchCleanup: {
+        value: null,
+        writable: true
+      },
+      _getUserMedia: {
+        value: options.getUserMedia,
+      },
+      _wasStopped: {
+        value: false,
+        writable: true
+      }
+    });
+
+    // Note(mpatwardhan) as a workaround for webkitbug208516(https://bugs.webkit.org/show_bug.cgi?id=208516)
+    // start watching document visibility and reacquire audio track if ended or muted.
+    if (options.workaroundWebKitBug1208516) {
+      this._startWatching(mediaStreamTrack);
+    }
+  }
+
+  _shouldAttemptRestart() {
+    const isDocumentVisible = document && document.visibilityState === 'visible';
+    return isDocumentVisible && (this.mediaStreamTrack.muted || this.mediaStreamTrack.readyState === 'ended');
+  }
+
+  _acquireTrack() {
+    const mediaStreamPromise = this._getUserMedia({ audio: true, video: false });
+    return mediaStreamPromise.then(mediaStream => {
+      return mediaStream.getAudioTracks()[0];
+    });
+  }
+
+  _handleTrackStateChange() {
+    if (this._shouldAttemptRestart()) {
+      // give it a moment before re-acquiring.
+      setTimeout(() => {
+        this._log.debug('checking if audio track needs to be replaced');
+        if (this._shouldAttemptRestart()) {
+          // replace tracks.
+          this._acquireTrack().then(newTrack => {
+            this._log.debug('replacing track, ', newTrack.id);
+            this._trackSender.replaceTrack(newTrack);
+            // if the track was not manually stopped,
+            if (!this._wasStopped) {
+              // start watching new track
+              this._startWatching(newTrack);
+            }
+          }).catch(err => {
+            this._log.debug('Failed to acquire track', err);
+          });
+        }
+      }, 10);
+    }
+  }
+
+  _startWatching(track) {
```

Comment:
@makarandp0 I think this can be simplified to:
```js
class LocalAudioTrack {
  _didCallStop: { value: false, writable: true }

  _restart(constraints) {
    return super._restart(constraints).then(() => {
      this._didCallStop = true;
      this._restartWhenInadvertentlyStopped();
    });
  }

  _restartWhenInadvertentlyStopped() {
    waitForEvent(this, 'stopped').then(() => {
      if (this._didCallStop) {
        return;
      }
      if (document.visibilityState === 'visible') {
        return this._restart();
      }
      return waitForEvent(document, 'visibilitychange').then(() => {
        // Is this necessary?
        return waitForSomeTime(10);
      }).then(() => {
        return this._restart();
      });
    });
  }
}
```

Reply:
@manjeshbhargav - I looked into this alternative, and found that track does not reliably get ended event. For example in this code 
```
var stream = await navigator.mediaDevices.getUserMedia({ audio: true, video: false});
var track = stream.getAudioTracks()[0];
track.addEventListener('ended', () => console.log('ended'));
track.stop();
assert(track.readyState === 'ended');
```
we do not get ended event. I think I will keep my version.

Reply:
@makarandp0 Let's discuss this today.

Reply:
I have addressed this by logging in the caller side.